duckduckgo/Android
# Implement favorites on Android

Task/Issue URL: https://app.asana.com/0/488551667048375/1200182978612539/f
Tech Design URL: 
CC: 

**Description**:
This PR introduces favorites in the Android app.
List of new features:
- Bookmark screen redesigned to show favorites
- Users can add favorites from the overflow menu
- Favicon logic improved to fetch high-res favicons
- Quick access grid on the home tab (and on the SystemSearchActivity)
- Quick access grid allow users to drag-drop and rearrange favorites (also edit and delete via long-press)
- autocomplete shows favorites as results
- Focusing the URL bar when browsing also show a grid with favorites


**Steps to test this PR**:
>> Fresh install
1. Fresh install
1. Visit any website and add it as favorite
2. During onboarding, favorites will not appear in the home tab if any Dax cta is active
3. Finish onboarding
4. Ensure favorites are shown in the home tab screen
5. Go to bookmarks screen, ensure favorites are listed

>> Re-arrange favorites
1. Add more favorites
2. Go to a new tab
3. Re-arrange favorites using drag and drop
4. Go to bookmarks screen
5. Ensure favorites list reflects previous changes (position)

>> Edit/Delete
1. You can edit/delete (and undo) favorites
2. Ensure those actions work from the home tab and from bookmarks screen

>> Autocomplete
1. Having multiple favorites
2. Go to URL bar and introduce any word part of favorites (title or URL)
3. ensure favorites appear in the suggestions
4. When browsing, focus the URL bar
5. Ensure favorites are shown
6. if you start typing, suggestions will appear instead
7. clearing the input text will show favorites again

>> Favicon
1. When adding websites as favorites, we know accept normal favicon (bitmap) and high-res favicons(apple-touch)
2. Visit tesla.com and add it as favorite
2. Ensure high-res favicon is persisted (previously, for this site, a low-res favicon was used)
3. For websites without favicon (or that we haven't still stored any favicon) ensure a new default favicon appears
4. Default favicon shows a letter and a random color

>> Database migration
1. Check out develop
2. run the app
2. add some bookmarks
3. Check out this branch
4. run the app
5. Ensure bookmarks are still present
6. Ensure migration did not crash


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
